shared/src/current_block.rs

```diff
@@ -0,0 +1,73 @@
+use anyhow::{anyhow, Context, Result};
+use futures::{stream::FusedStream, Stream};
+use primitive_types::H256;
+use std::time::Duration;
+use web3::{
+    types::{BlockId, BlockNumber},
+    Transport, Web3,
+};
+
+pub type Block = web3::types::Block<H256>;
+
+const POLL_INTERVAL: Duration = Duration::from_secs(1);
+
+/// A stream that yields the current block whenever it changes.
+pub fn block_stream(web3: Web3<impl Transport>) -> impl Stream<Item = Block> + FusedStream {
+    // We do not use a block filter because filters have been unreliable in the past, do not work
+    // with load balancing over multiple nodes and would still need to be polled unless we used web
+    // sockets.
+    futures::stream::unfold(H256::zero(), move |previous| {
+        let web3 = web3.clone();
+        async move {
+            loop {
+                tokio::time::delay_for(POLL_INTERVAL).await;
```

Comment:
nit: shouldn't we delay at the end of the loop to yield the current block without delay?

Reply:
Can't really have it just at the end because want to yield the current block from the stream immediately, but that returns from the function. We could keep it at the start but have a bool with which we can skip the very first sleep.  Adding the bool is only  3 lines but on the other hand I don't think it's wrong to sleep immediately either.